src/mlpack/methods/kde/kde_rules_impl.hpp

```diff
@@ -0,0 +1,194 @@
+{
+  double distance = metric.Evaluate(querySet.col(queryIndex),
+                                    referenceSet.col(referenceIndex));
+  if (tree::TreeTraits<TreeType>::RearrangesDataset)
+    densities(oldFromNewQueries.at(queryIndex)) += kernel.Evaluate(distance);
+  else
+    densities(queryIndex) += kernel.Evaluate(distance);
+  ++baseCases;
+  lastQueryIndex = queryIndex;
+  lastReferenceIndex = referenceIndex;
+  return distance;
+}
+
+//! Single-tree scoring function.
+template<typename MetricType, typename KernelType, typename TreeType>
+double KDERules<MetricType, KernelType, TreeType>::
+Score(const size_t /* queryIndex */, TreeType& /* referenceNode */)
+{
+  ++scores;
+  traversalInfo.LastScore() = 0.0;
+  return 0.0;
+}
+
+template<typename MetricType, typename KernelType, typename TreeType>
+double KDERules<MetricType, KernelType, TreeType>::Rescore(
+    const size_t /* queryIndex */,
+    TreeType& /* referenceNode */,
+    const double oldScore) const
+{
+  // If it's pruned it continues to be pruned.
+  return oldScore;
+}
+
+//! Double-tree scoring function.
+template<typename MetricType, typename KernelType, typename TreeType>
+inline double KDERules<MetricType, KernelType, TreeType>::
+Score(TreeType& queryNode, TreeType& referenceNode)
+{
+  const double maxKernel =
+    kernel.Evaluate(queryNode.MinDistance(referenceNode));
+  const double minKernel =
+    kernel.Evaluate(queryNode.MaxDistance(referenceNode));
+  const double bound = maxKernel - minKernel;
+  double score;
+
+  if (bound <= (absError + relError * minKernel) / referenceSet.n_cols)
+  {
+    // Auxiliary variables.
+    double kernelValue;
+    arma::vec& referenceCenter = referenceNode.Stat().Centroid();
+    arma::vec& queryCenter = queryNode.Stat().Centroid();
+
+    // If calculating a center is not required.
+    if (tree::TreeTraits<TreeType>::FirstPointIsCentroid)
+    {
+      kernelValue = EvaluateKernel(queryNode.Point(0), referenceNode.Point(0));
+    }
+    // If a child center is the same as its parent center.
+    else if (tree::TreeTraits<TreeType>::HasSelfChildren)
+    {
+      // Reference node.
+      if (referenceNode.Parent() != NULL &&
+          referenceNode.Point(0) == referenceNode.Parent()->Point(0))
+        referenceCenter = referenceNode.Parent()->Stat().Centroid();
+      else
+      {
+        referenceNode.Center(referenceCenter);
+      }
+      // Query node.
+      if (queryNode.Parent() != NULL &&
+          queryNode.Point(0) == queryNode.Parent()->Point(0))
+        queryCenter = queryNode.Parent()->Stat().Centroid();
+      else
+      {
+        queryNode.Center(queryCenter);
+      }
+      // Compute kernel value.
+      kernelValue = EvaluateKernel(queryCenter, referenceCenter);
+    }
+    // Regular case.
+    else
+    {
+      referenceNode.Center(referenceCenter);
+      queryNode.Center(queryCenter);
+      kernelValue = EvaluateKernel(queryCenter, referenceCenter);
+    }
+
+    #pragma omp for
```

Comment:
Did you see good speedup with this?  Just curious.  I have played with using OpenMP at higher levels of the tree traversal and it seemed to work ok, so it would be interesting to know if it helps here or if the overhead of giving all the threads work is too much here.

Reply:
I've been playing with this and times are pretty much the same, can't see any big difference actually.

After executing a few quick tests and adding up the times for KDE computation I get this results:

* Without OMP: 369.532391s
* With OMP: 364.340302s

Almost in all tests the OMP execution gives a very slightly better result than the one without OMP, but this difference is so small that I'm not sure if this is enough to justify using OMP.